File: 57_insert_interval/insert.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> result = new ArrayList<>();

        for (int[] interval : intervals) {
            if (interval[1] < newInterval[0]) {
                result.add(interval);
            } else if (interval[0] > newInterval[1]) {
                result.add(newInterval);
                newInterval = interval;
            } else {
                int start = Math.min(interval[0], newInterval[0]);
                int end = Math.max(interval[1], newInterval[1]);
                newInterval = new int[] {start, end};
            }
        }

        result.add(newInterval);

        return result.toArray(new int[result.size()][]);
    }
}

File: 57_insert_interval/insert.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> res;

        for (int i = 0; i < intervals.size(); i++) {
            vector<int> interval = intervals[i];

            if (newInterval[1] < interval[0]) {
                res.push_back(newInterval);
                res.insert(res.end(), intervals.begin() + i, intervals.end());
                return res;
            } else if (newInterval[0] > interval[1]) {
                res.push_back(interval);
            } else {
                int start = min(newInterval[0], interval[0]);
                int end = max(newInterval[1], interval[1]);
                newInterval = {start, end};
            }
        }

        res.push_back(newInterval);

        return res;
    }
};